[1단계 - 장기] 강산(조재중) 미션 제출합니다.

m-a-king opened the pull request:
> <!-- 
> 
> ## 코드 리뷰 팁
> 
> - 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
>   - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)
> 
> -->
> 
> ## 체크 리스트
> 
> - [x] 미션의 필수 요구사항을 모두 구현했나요?
> - [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
> - [x] 애플리케이션이 정상적으로 실행되나요?
> - [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
>   - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
>   - 
> https://prolog.techcourse.co.kr/studylogs/4107
> 
> ## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
> 
> ### 1~5점 중에서 선택해주세요.
> 
> - [ ] 1 (전혀 충족하지 못함)
> - [ ] 2
> - [ ] 3 (보통)
> - [x] 4
> - [ ] 5 (완벽하게 충족)
> 
> ### 선택한 점수의 이유를 적어주세요.
> 
> 코드를 적절히 리팩토링하여 중복을 줄였으나, 아직 리팩토링 과정에서 고민해야 할 점을 인지하고 있습니다. 원시 값 포장을 통한 값 객체와 일급 컬렉션을 적극적으로 활용할 예정입니다.
>   
> ## 어떤 부분에 집중하여 리뷰해야 할까요?
> 
> 안녕하세요, 베루스!
> 강산입니다. 
> 
> > 강산은 제 고양이 이름입니다.
> 
> 1 레벨의 마지막 미션, 장기 리뷰를 맡아주셔서 감사합니다.
> 베루스의 사소한 피드백부터 깊고 주관적인 견해까지 너무너무 기대됩니다.
> 열심히 소통하며 배우고 싶어요!
> 
> ---
> 
> <!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
> 예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->
> 
> ## Position 외 추가 포장(Distance, Row/Column)
> 
> 현재 Position 클래스 외에도 거리나 행과 열의 개념을 명확히 표현하기 위해 추가적으로 Distance, Row/Column 을 값 객체나 ENUM으로 활용하여 개념을 포장하고자 합니다. 이 방식에 대한 더 나은 아이디어가 있는지 여쭤보고 싶습니다.

안녕하세요, 베루스!

양질의 피드백 너무 감사합니다. 덕분에 제 생각을 많이 정리해본 것 같습니다.

제가 작성한 여러 MoveStrategy 구현체들 중 일부는 상태가 없는 객체입니다. 이런 상태가 없는 객체에 대해 크루들과 코치님들께서 다양한 피드백을 주셨는데요, 이를 바탕으로 상태가 없더라도 정상적인 객체일 수 있다고 생각하는 저의 의견을 정리해봤습니다. 
베루스에게도 피드백 받을 수 있을까요?

# 의견
상태가 없어도 객체일 수 있다는 관점은 객체지향 설계의 본질인 행위 중심적 사고를 잘 드러낸다. 흔히 객체는 데이터와 행위를 함께 묶어 캡슐화한 존재라고 설명되지만, 실제로 객체지향적 설계가 강조하는 바는 데이터 자체보다는 그 데이터를 다루는 행위에 있다.
이런 맥락에서 상태가 없는 객체란 데이터를 전혀 가지지 않거나 내부적으로 유지하지 않는 객체를 의미한다. 예를 들어 전략/정책/유틸리티 객체가 이런 상태 없는 객체에 해당한다. 상태가 없는 객체는 자신에게 주어진 특정 입력을 기반으로 정의된 행위만을 수행하고, 입력에 따라 항상 같은 결과를 만들어내는 순수한 형태의 객체다.

상태가 없는 객체가 유효한 객체로 인정받을 수 있는 이유는 객체가 본래 데이터(상태)와 행위(메서드)의 묶음이라는 정의보다 더 넓은 의미를 갖기 때문이다. 캡슐화란 외부 세계와의 구체적 의존성을 숨기고 필요한 행동만 외부에 노출하여 객체가 가진 책임을 명확히 하는 것을 의미한다. 따라서 어떤 객체가 내부 상태를 유지하지 않더라도, 그 객체가 외부로부터 입력받은 데이터에 대해 명확한 행위를 수행할 책임과 목적을 지닌다면 이는 충분히 캡슐화된 객체로 볼 수 있지 않은가?

### Q. 그렇다면 상태가 없는 객체의 응집도와 결합도는 어떻게 평가되는가?
응집도는 객체가 지닌 책임의 일관성 또는 목적의 명확성을 나타낸다. 상태가 없는 객체는 본래 특정한 목적을 위해 명확한 행위만 수행하도록 설계된 경우가 많으므로 일반적으로 높은 응집도를 가진다. 예를 들어 StraightMoveStrategy라는 전략 객체가 “장기말의 직선 이동 검증”이라는 단 하나의 책임만을 수행하고 있다면, 이는 명확한 단일 책임으로 높은 응집도를 가진다.

결합도는 한 객체가 다른 객체에 얼마나 의존하는지를 나타낸다. 상태가 없는 객체는 일반적으로 외부에서 주어진 데이터에만 의존하며, 내부에 별도의 상태를 가지지 않기 때문에 다른 객체와의 강한 연결고리를 형성할 필요가 없다. 만약 이 객체가 인터페이스 기반으로 정의되어 있고, 특정 구현체를 선택적으로 교체하여 사용할 수 있도록 설계되어 있다면 객체 간 결합도는 더욱 낮아진다.

이렇게 상태가 없는 객체는 객체지향 설계에서 불변성을 유지하고, 객체 간 의존성을 최소화하며, 테스트 용이성을 높이는 데 기여할 수 있다. 반면, 객체의 상태나 과거의 행위에 따라 로직이 복잡하게 변해야 하는 경우에는 상태가 없는 객체만으로 이를 해결하기는 어렵다. 따라서 상태가 없는 객체의 사용 여부는 요구사항과 문제의 성격에 따라 결정되어야 한다.

예를 들어 MoveStrategy 인터페이스를 다형적으로 확장하여 StraightMoveStrategy, CurvedMoveStrategy, SingleMoveStrategy 등 다양한 이동 전략을 정의할 수 있다. 이때 구현되는 전략 객체들은 상태가 없다면 런타임에 더 간편하게 교체할 수 있고, 불변성이나 테스트 용이성 측면에서도 장점이 많아진다. 하지만 요구사항에 따라 특정 전략이 내부에 상태를 유지해야 하는 상황도 충분히 존재할 수 있다. 이 경우 상태를 가진 전략 객체는 교체나 테스트에 있어 상태 없는 객체보다는 약간 더 복잡해질 수 있지만, 여전히 전략 패턴의 본질적인 장점 (확장성, 다형성)을 잃지는 않는다.

결국, 상태가 있는 전략 객체와 상태가 없는 전략 객체는 우열 관계가 아니라 요구사항과 설계의 목적에 따라 결정할 수 있다. 상태를 갖는 객체가 필요하다면 이를 자연스럽게 설계에 포함시키되, 객체가 가진 상태가 외부로 불필요하게 노출되지 않도록 캡슐화의 원칙을 지켜 관리하는 것이 중요하다.

### Q. 상태가 없이 메서드만 있는 클래스의 경우 static 메서드만 가진 클래스로 바꿀 수 있지 않은가?
둘이 기능적으로 비슷해 보이지만 객체지향 관점에서는 다르다고 생각합니다. 비록 상태가 없고 메서드만 있더라도, 그 클래스가 인스턴스화된 객체의 메서드로 동작하는지, 아니면 클래스 수준의 static 메서드로서 동작하는지는 설계적 차이를 만들어낼 수 있습니다.
객체지향에서 메서드는 인스턴스에게 메시지를 보내 그 객체가 해당 메시지에 책임을 가지고 응답하도록 하는 방식으로 동작합니다. 즉, 인스턴스 메서드를 호출한다는 건, 객체에게 질문을 던지는 것이며 객체는 자신이 책임을 지고 있는 행위를 통해 응답합니다.
반면 static 메서드는 인스턴스에게 묻는 것이 아니라, 이미 클래스에 고정적으로 기재된 방법을 직접 호출하는 것이라고 생각합니다.

### Q. 인스턴스 메서드를 쓰는 것과 static 메서드를 쓰는 것은 (객체 지향 관점에서) 어떤 차이점이 있어서 전자는 객체이고 후자는 객체가 아니라고 생각할 수 있을까?
static 메서드는 행위를 외부에서 주입할 수도 없고, 동적으로 전략을 바꾸거나 다형적으로 대체하기도 어렵습니다.
따라서 static 메서드로의 전환은 단순히 구현 방식이 다를 뿐 아니라, 객체지향 설계의 본질 중 일부를 흐릴 수 있다고 생각합니다. (static 메서드는 절차지향적 접근과 유사해집니다. “클래스”라는 이름의 이름공간에 함수를 묶어둔 형태로, 호출하는 순간 이미 특정 구현체를 직접 써야 하므로 유연성이 떨어집니다.)
요약: 인스턴스 메서드는 메시지를 받을 주체가 명확하여 객체 간 협력과 다형적 교체가 가능하지만, static 메서드는 호출 대상이 고정된 절차적 접근에 가깝다.

---

긴 글 읽어주셔서 감사합니다.

jhan0121 committed:
docs: 1.1단계 - 보드 초기화 요구 기능 사항 정의

jhan0121 committed:
docs: 1.1단계 - 보드 초기화 요구 기능 사항 정의

jhan0121 committed:
feat: Position 추가

jhan0121 committed:
Merge remote-tracking branch 'origin/step1' into step1

# Conflicts:
#	README.md

m-a-king committed:
feat: 장기 말 추가

jhan0121 committed:
feat: Board 추가

jhan0121 committed:
fix: 이동 가능 지점 판정 기능 수정

m-a-king committed:
feat: Chariot(차) 이동 검증 로직 추가

jhan0121 committed:
feat: General(초,한), Guard(사) 추가

jhan0121 committed:
feat: Horse(마), Elephant(상) 추가

m-a-king committed:
feat: Cannon 이동 검증 로직 추가

m-a-king committed:
feat: 장기 말 초기 배치 기능 추가

jhan0121 committed:
feat: Position 캐시 적용

jhan0121 committed:
feat: Position 행/열 차이 계산 기능 추가

jhan0121 committed:
feat: 경로 계산 기능 추가

m-a-king committed:
feat: 장기 말 이동 구현, 이동 규칙 검증 추가

jhan0121 committed:
docs: 장기 기물 요구 기능 사항 추가

jhan0121 committed:
fix: Route 이동 경로 산정 기능 수정

jhan0121 committed:
refactor: 불필요한 클래스 삭제

jhan0121 committed:
refactor: 초기 위치 제공 기능 수정

jhan0121 committed:
refactor: Board 기물 이동 기능 개선

jhan0121 committed:
refactor: Position 코드 스타일 개선

jhan0121 committed:
refactor: Route 코드 스타일 개선

jhan0121 committed:
refactor: Team 코드 스타일 개선

jhan0121 committed:
fix: 테스트 실패 해결

- board: move 메서드명 변경
- position: 에러 메시지 변경
- Route: 도착지 제거 반영

jhan0121 committed:
test: Movement 테스트 추가

m-a-king committed:
refactor: import 구조 변경

m-a-king committed:
test: Board 테스트 추가

m-a-king committed:
test: Soldier 테스트 추가

m-a-king committed:
test: Horse 테스트 추가

m-a-king committed:
test: Guard 테스트 추가

m-a-king committed:
test: General 테스트 추가

m-a-king committed:
test: Elephant 테스트 추가

jhan0121 committed:
test: 테스트 코드 개선

jhan0121 committed:
refactor: 규칙 이동 준수 여부 기능 개선

jhan0121 committed:
test: Chariot 테스트 추가

jhan0121 committed:
test: Cannon 테스트 추가

jhan0121 committed:
test: Piece 테스트 추가

- 포가 아닌 기물에 대한 경로에 기물 존재 여부에 따른 테스트
- 포가 아닌 기물에 대한 목적지에 아군 존재 여부에 따른 테스트

m-a-king committed:
refactor: 패키지 이동

m-a-king committed:
feat: 프로그램 실행 기능 추가 (보드판 초기화)

m-a-king committed:
test: 팀 기준 행 결정 테스트 추가

m-a-king committed:
refactor: 리스코프 치환 원칙 준수를 위해 인터페이스 기반 구성(Composition)으로 전환

- 사소한 로직 개선 포함.

verus-j commented on src/main/java/janggi/piece/Cannon.java:
```diff
@@ -0,0 +1,57 @@
+        return new Cannon(position, team, new StraightMoveStrategy(), new RequiredBlockCountStrategy(REQUIRE_BLOCK_COUNT));
+    }
+
+    public static List<Cannon> Default(Team team) {
```
메서드 명을 Default로 해도 될까요? [여기](https://tecoble.techcourse.co.kr/post/2020-04-26-Method-Naming/)를 참고하면 좋을 것 같네요! 그리고 모든 기물의 Default 메서드의 테스트가 없네요. default일 경우 초기 위치는 충분히 테스트할만하지 않을까요?

m-a-king replied:
> 메서드 명을 Default로 해도 될까요? [여기](https://tecoble.techcourse.co.kr/post/2020-04-26-Method-Naming/)를 참고하면 좋을 것 같네요!

감사합니다! 메서드 명이 완전히 잘못된 것 같습니다. 수정하겠습니다.

> 그리고 모든 기물의 Default 메서드의 테스트가 없네요. default일 경우 초기 위치는 충분히 테스트할만하지 않을까요?

향후 "기물 배치 규칙이 `상마상마 마상마상`과 같이 여러 가지로 분기" 되거나 "초기 위치가 동적으로 결정"되는 구조로 확장될 경우엔 반드시 테스트가 필요하다고 생각하고 있고, 그 시점에 테스트 추가를 고려하고 있습니다.

혹시 현재 구조에서도 이 부분을 명시적으로 테스트하는 것이 좋겠다고 판단하신 이유가 있는지 여쭤보고 싶습니다!

verus-j replied:
상마상마, 마상마상과 같은 기물 배치 규칙이 추가되었을 때만 특별히 테스트를 추가하는 이유가 있을까요?

m-a-king replied:
public한 상수가 있다면, “상수가 맞는지 테스트하는 것”과 비슷한 맥락으로 느껴졌습니다.
예를 들어, 3이라는 값을 작성하고 테스트에서 그게 정말 3이 맞는지를 확인하는 식이라, 마치 게터를 테스트하는 것처럼 조금은 당연한 내용을 검증하는 느낌이 들어 처음에는 테스트의 필요성에 의문이 들었습니다. 
하지만 블랙잭을 예시로, 블랙잭 상수가 21이 맞는지를 테스트하는 것도 의미가 있을 수 있다고 생각이 들기도 합니다.

물론 이 경우는 그보다는 더 복잡한 로직이고, 실수가 있을 수 있습니다. 그런 점에서 장기의 초기 세팅 역시 테스트할 충분한 가치가 있는 것 같습니다.

정리하자면, 처음에는 꼭 테스트가 필요할까? 하는 생각이 있었지만, 테스트를 추가해서 나쁠 건 없고, 명시적으로 확인해두는 것도 좋은 방향일 수 있겠다는 생각이 들었습니다. 다만 개인적으로는 약간 효율 면에서 고민이 되는 테스트라는 인상도 있었습니다.

> 상마상마, 마상마상과 같은 기물 배치 규칙이 추가되었을 때만 특별히 테스트를 추가하는 이유가 있을까요?

초기 상차림이 여러 종류로 확장된다면, 테스트를 추가할 이유는 충분하다고 생각합니다. 선택한 규칙에 따라 올바르게 세팅되고 있는지를 명확히 확인할 수 있다는 점에서 의미 있는 테스트라고 생각합니다!

verus-j replied:
초기 세팅에 대한 테스트가 있으면, 후에 관련 로직의 변경이 있을 때 인지를 할 수 있게 됩니다. TDD로 진행하게 된다면 원하는 초기 세팅에 맞는 테스트로 변경 후 구현을 진행하여 안정성을 확보할 수 있겠네요. 혹여나 잘못된 코드로 수정한다면 테스트 실패로 인해 피드백을 받고 원하는 방향으로 수정할 수 있습니다. 이외에도 테스트 코드를 통해 문서화 효과도 볼 수 있을 것 같네요. 이러한 이유 때문에 기물의 초기 세팅 테스트를 추가하는걸 권했습니다~

m-a-king replied:
감사합니다. 역시 테스트는 정말 중요한 것 같아요!

verus-j commented on src/main/java/janggi/piece/Chariot.java:
```diff
@@ -0,0 +1,52 @@
+        return Chariot.of(position, team);
+    }
+
+    @Override
+    protected void validateSpecialRule(final Board board, final Position destination) {
+        // 특수 규칙 없음
+    }
```
hook 메서드를 아실까요? 참고해서 개선해보면 좋을 것 같네요~

m-a-king replied:
네, 상위 클래스에서 바디가 비어있는 메서드를 선언하고 이를 필요한 하위 클래스에서만 오버라이딩을 통해서 사용하는 방법이라고 알고 있습니다.

특수 규칙이 없다는 점 자체가 해당 클래스의 중요한 특징이기 때문에, 상위 클래스에 비어 있는 hook 메서드를 두고 필요한 곳에서만 사용하기보다, 하위 클래스가 명시적으로 이를 구현하도록 강제하는 추상 메서드 형태를 선택했습니다.
이렇게 하면, 특수 규칙이 없음을 구현의 명시성과 의도를 통해 드러낼 수 있어, 코드 가독성과 유지보수에 유리하다고 판단했습니다.

verus-j replied:
`// 특수 규칙 없음`이라는 주석이 필요한걸까요? hook 메서드를 두면 충분히 주석을 대체할 수 있는 것 같습니다. 그리고, 구현이 없는 하위 타입이 많아서 코드를 파악하기 더 어려운 것 같네요... 오버라이드한 구현체를 IDE로 조회해보면 대부분 빈 구현이더라구요. 필요한 몇몇 타입만 오버라이드하면 더 쉽게 파악할 것 같아요.

m-a-king replied:
다시 한 번 명확히 짚어주셔서 감사합니다!

말씀해주신 대로 hook 메서드를 사용하면 공통 템플릿 구조가 깔끔해지고, 구현하지 않는 타입이 많을 경우 오히려 가독성이 좋아질 수 있다는 점에 공감합니다. 실제로 피드백을 받고 즉시 적용해보기도 했습니다.

다만 제 입장에서는 특수 규칙이 없다는 점 자체가 해당 클래스의 중요한 특징이라고 생각해서, 이를 명시적으로 표현하고 싶다는 의도가 있었습니다.
hook 메서드의 경우, 구현이 없다는 점을 보고 규칙이 없음을 암묵적으로 해석해야 하는 반면,
추상 메서드를 구현한 뒤 // 특수 규칙 없음이라고 명시해두면, 의도적으로 비워두었음을 코드상에서 더 분명히 드러낼 수 있다고 판단했습니다.

> 오버라이드한 구현체를 IDE로 조회해보면 대부분 빈 구현이더라구요

말씀 주신 것처럼, 구현이 없는 하위 클래스가 많아질 경우 IDE 상에서 오히려 파악이 어려워질 수 있다는 점은 미처 고려하지 못했던 부분이었습니다.
말씀해주신 방향대로 훅 메서드 형태로 수정하겠습니다. 이후 특수 규칙이 다양해지는 시점에 추상 메서드로 다시 전환하는 것도 고려해보겠습니다!

verus-j commented on src/main/java/janggi/Board.java:
```diff
@@ -0,0 +1,57 @@
+        return positionToPiece.containsKey(position);
+    }
+
+    public void movePiece(final Position departure, final Position destination) {
+        Piece allyPiece = getPiece(departure);
+        Piece movedPiece = allyPiece.move(this, destination);
+
+        Score score = positionToPiece.remove(departure).die();
+        // TODO 2.1 단계(점수 계산) 요구사항 추가 예정
+        updateBoard(destination, movedPiece);
+    }
```
사용하지 않는 메서드네요!

m-a-king replied:
네, 테스트가 지워지면서 사용하지 않는 메서드로 남은 것 같습니다. 이후 2.1단계부터 구현 예정입니다.

verus-j commented on src/main/java/janggi/Board.java:
```diff
@@ -0,0 +1,57 @@
+        throw new IllegalArgumentException("장기말이 존재하지 않는 지점입니다.");
+    }
+
+    public Map<Position, Piece> getPositionToPiece() {
+        return Collections.unmodifiableMap(positionToPiece);
+    }
```
여기도 사용하지 않는 메서드네요

m-a-king replied:
네, 테스트가 지워지면서 사용하지 않는 메서드로 남은 것 같습니다. 이후 2.1단계부터 구현 예정입니다. 뷰에서 사용할 것 같습니다.

verus-j commented on src/main/java/janggi/Team.java:
```diff
@@ -0,0 +1,23 @@
+
+    public static int decideRow(final int row, final Team team) {
+        if (team.isGreen()) {
+            int rowFlipBase = 11;
```
rowFlipBase는 어떤 의미일까요?? 변수명이 조금 애매하다고 느껴져서 11이라는 값이 어떤 의미인지 쉽게 알 수 없을 것 같습니다!

m-a-king replied:
맞습니다. 고민을 오래 했던 변수명이지만 아직 명확하지 않다고 느껴집니다. 

각 팀의 같은 종류의 장기 말의 초기 위치의 합이 11인 점을 이용해서 중복된 코드를 없애려고 했습니다.

이런 계산 없이 하나씩 직접 넣어주는 편이 좋았을까요?

verus-j replied:
> 각 팀의 같은 종류의 장기 말의 초기 위치의 합이 11인 점을 이용해서 중복된 코드를 없애려고 했습니다.

중앙선이 필요했던거군요. 적절한 이름만 찾을 수 있다면 위의 방식이 좋을 것 같긴합니다. 조금 더 좋은 이름을 고민해보면 좋을 것 같아요.

m-a-king replied:
네 맞습니다.

> 각 팀의 같은 종류의 장기 말의 초기 위치의 합이 11인 점을 이용해서 중복된 코드를 없애려고 했습니다.

각 팀의 같은 종류의 장기 말의 초기 행의 합이 11인 점을 이용해서 중복된 코드를 없애려고 했습니다.

행으로 말씀드려야 하는데 위치라고 표현했네요. 죄송합니다.
11이라는 수가 장기와는 관련이 없어서 변수명을 명확히 짓기가 곤란했던 것 같습니다. 

혹시 sumOfRow, rowTotal 이 더 명확했을까요? 너무 복잡하게 생각했을까요?

verus-j replied:
차라리 sumOfRow가 더 의미가 맞을 것 같네요. 더 풀어 쓰면 sumOfEachTeamPieceRow가 나올 것 같은데, 조금 긴 것 같아 고민이 되네요. 아무튼, 각 팀 장기 말의 초기 행의 합이라는 의미를 주는 변수명으로 수정하는게 좋을 것 같습니다~

m-a-king replied:
넵. 더 풀어서 제안해주신 sumOfEachTeamPieceRow가 명확하다고 생각해서 이것으로 수정하겠습니다. 감사합니다!

verus-j commented on src/main/java/janggi/Board.java:
```diff
@@ -0,0 +1,57 @@
+
+    private final Map<Position, Piece> positionToPiece;
+
+    private Board(final Map<Position, Piece> positionToPiece) {
+        this.positionToPiece = positionToPiece;
+    }
+
+    public static Board initialize(final List<Piece> pieces) {
+        HashMap<Position, Piece> positionToPiece = new HashMap<>();
+
+        pieces.forEach(piece -> positionToPiece.put(piece.getPosition(), piece));
+
+        return new Board(positionToPiece);
+    }
```
initialize라는 이름의 정적 팩터리 메서드를 만들 필요가 있을까요? 생성자를 활용하면 안되나요? 이미 생성자는 initialize라는 의미를 가지고 있습니다.

m-a-king replied:
우선 initialize()라는 메서드명은 지금 상황에서 적절하지 않아 보입니다.
이전에 내부에서 기물을 직접 생성했기 때문에 초기화를 의미하는 네이밍이 어울렸을 수 있지만, 현재처럼 List<Piece> pieces를 외부에서 주입받는 구조에서는 적절한 이름이 아닙니다.

현재 역할은 “외부에서 주어진 기물 목록으로 객체를 생성”하는 것이므로, 생성자를 사용해도 충분히 의도를 드러낼 수 있고, 정적 팩터리 메서드를 굳이 사용할 필요는 없어 보입니다.

하지만 현재 대부분의 클래스가 정적 팩터리 메서드를 사용하고 있고, 그 흐름 안에서 이 클래스만 new를 사용하는 건 일관성 측면에서 오히려 눈에 띄고 다르게 느껴질 수 있습니다.

따라서 생성자 대신 from(List<Piece> pieces) 같은 정적 팩터리 메서드를 사용하는 방향도 고려해보는 것은 어떤지 여쭤보고 싶습니다. 특별한 이유가 없다면 전체 구조와 컨벤션에 맞추는 게 더 낫다고 판단했습니다!

verus-j commented on src/main/java/janggi/Board.java:
```diff
@@ -0,0 +1,57 @@
+        Piece allyPiece = getPiece(departure);
+        Piece movedPiece = allyPiece.move(this, destination);
+
+        Score score = positionToPiece.remove(departure).die();
```
출발지의 기물을 왜 die 시킬까요?

m-a-king replied:
스코어 계산을 이렇게 할 것 같다고 페어와 얘기하면서 테스트도 없이 작성하다 보니 실수한 것 같습니다...

34번 줄의 departure를 destination으로 변경하고, updateBoard 메서드에서 departure를 remove할 것 같습니다.

verus-j commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public Piece move(Board board, Position destination) {
+        validateMove(board, destination);
```
안에 구조가 복잡해서 가능할지 모르겠는데, validateMove 로직을 Board에서 실행하면 안되나요?? Piece가 Board에 대해 꼭 의존을 해야하는지 의문이 생겨 여쭤봅니다.

m-a-king replied:
맞습니다! 구조적으로는 기물(Piece)이 보드(Board)에게 자신이 움직이기로 결정한 루트를 제공하며 Board에서 검증하는 것이 더 좋은 것 같습니다.

이러한 코드가 만들어진 이유는 기물이 주체적인 객체로 스스로 판단하고 행동하게 하고 싶습니다.
그래서 보드는 단순히 기물이 주변 상황을 파악하기 위한 수단으로만 제공됩니다. 
(Map을 사용하지 않았다면, 이름이 보드가 아니라 Pieces가 될 수도 있을 것 같습니다)

기물 스스로 갈 길을 판단하고, 그 과정에서 필요한 정보만 보드에서 조회하는 방식이 더 자연스럽게 느껴졌습니다.
다만, 보드를 기물에게 넘기는 구조가 순환참조처럼 보일 수 있어서 처음엔 어색했지만,
기물 중심의 철학이라면 이 정도의 의존성은 허용 가능한 수준이라 판단했습니다.

호출한 메서드에서 자신을 호출한 메서드를 다시 호출한다면, 스택오버플로우의 위험이 있는 것을 인지하고 있습니다. 
설계가 올바르지 못했는지 여쭤보고 싶습니다.

verus-j replied:
스택오버플로우의 위험은 테스트 코드를 작성하면 방지할 수 있는 문제입니다. 순환참조가 무조건 나쁜건 아니기에 필요하다면 사용해도 된다고 생각합니다. 그럼에도 순환참조가 코드의 구조를 복잡하게 보이는 경향이 있어 의견을 드려봤습니다~

m-a-king replied:
> 스택오버플로우의 위험은 테스트 코드를 작성하면 방지할 수 있는 문제입니다. 

감사합니다. 명확히 해결할 수 있다는 걸 미처 생각 못 했네요. 말씀 듣고 보니 당연한데 놓치고 있었습니다.

---

처음에는 Piece에서 사용해야 하는 메서드만 함수형 인터페이스로 전달해볼까 고민했었는데요,
차라리 필요한 메서드들만 모은 인터페이스를 따로 분리하고, Board가 그 인터페이스를 구현하는 방식은 어떨까 하는 생각이 들었습니다.
이렇게 하면 Piece는 필요한 기능에만 의존하게 되니 좋은 것 같지만, `스택오버플로우의 위험은 테스트 코드를 작성하면 방지할 수 있는 문제입니다.` 라는 피드백을 보니 불필요한 복잡성이 또 추가되나 싶어서 여쭤봅니다!

``` java
public interface BoardContext{
    // Piece가 사용하는 최소한의 기능
}
```

``` java
public class Board implements BoardContext{
    ...
}
```

```java
    // 현재와 동일
    public void movePiece(final Player player, final Position departure, final Position destination) {
        ...
        Piece movedPiece = selectedPiece.move(this, destination);
       ...
    }
```

```java
    public Piece move(BoardContext boardContext, Position destination) {
        validateMove(boardContext, destination);
        return createPiece(destination);
    }
```

verus-j replied:
move에서만 사용할 인터페이스를 추출하는것도 좋은 방법이 될 수 있지만, 강산이 우려하는 것처럼 현재 상황에서는 불필요한 복잡성이 추가될 것 같네요 ㅠㅠ 지금처럼 구성하는 것도 괜찮다고 생각합니다.

m-a-king replied:
감사합니다. 많이 배웠습니다!

verus-j commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private void validateIsAlly(final Board board, final Position destination) {
+        if ((board.isExists(destination) && board.isAlly(destination, team))) {
```
isAlly에는 isExists의 의미도 같이 있는 것 같습니다. 따로 보지 말고 isAlly에서 isExists를 확인하는 로직도 같이 포함하는게 어떤가요?

m-a-king replied:
훨씬 이해하기 좋은 것 같습니다. 수정하겠습니다. 감사합니다!

verus-j commented on src/main/java/janggi/piece/strategy/block/BlockStrategy.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface BlockStrategy {
+
+    String exceptionMessage = "이동 경로에 기물 갯수가 조건에 맞지 않습니다.";
```
꼭 여기 있어야할 필드인가요?

m-a-king replied:
피드백을 받기 전까지는 해당 인터페이스의 모든 구현체가 `이동 경로에 기물 갯수가 조건에 맞지 않습니다.` 메시지를 동일하게 사용할 것이라고 생각했습니다.
실제로 구현을 고려했던 전략들도 대부분 `2개 이상 4개 미만의 기물이 있어야 한다, 홀/짝수 개의 기물이 있어야 한다` 등 기물의 수에 따른 조건 검증이었기 때문에, 메시지를 인터페이스에서 공통으로 관리하는 것이 적절하다고 판단했습니다.

하지만 포(Cannon)의 도메인 규칙처럼 `같은 종류의 기물(포)을 넘지 못한다`와 같이 종류 기반의 차단 조건을 구현체로 추가한다면, 현재 인터페이스에 선언된 메시지는 의미가 맞지 않거나 오히려 제약으로 작용할 수 있을 것 같다고 생각했습니다.

물론 아직 해당 전략을 구현할 계획은 없지만, 인터페이스에 정의된 예외 메시지가 향후 구현체의 범위를 제한할 수 있다는 점에서, 지금 구조가 정말 확장에 유연한지 고민이 생기기도 합니다.
이 상황에서, 현재처럼 인터페이스에 메시지를 정의해두는 것이 적절한 선택인지 의견을 여쭤보고 싶습니다.

verus-j replied:
* 예외 메시지의 중복을 꼭 제거해야할까요? 
* 각 케이스는 현재는 동일한 상황이지만, 언젠가는 개별 케이스로 변화할 수도 있지 않을까요? 
* 예외 메시지의 중복으로 인해 한쪽의 변경으로 다른 쪽의 수정을 놓치는 케이스도 발생할까요? 즉, 같은 이유로 변경할 다른 쪽의 예외 메시지가 있을까요?

이러한 점들을 고민해보면 좋을 것 같습니다.

m-a-king replied:
> 예외 메시지의 중복을 꼭 제거해야할까요?

예외 메시지의 중복이 있더라도, 심각한 문제가 발생하지는 않을 것으로 생각합니다. 그저 중복된 문자열이 존재하는 상황입니다.

> 각 케이스는 현재는 동일한 상황이지만, 언젠가는 개별 케이스로 변화할 수도 있지 않을까요?

맞습니다! 말씀하신 것처럼 현재는 동일한 상황이지만, 앞으로는 개별적으로 변화할 가능성이 충분히 있습니다. 다만, 개별적인 변화가 필요해지면 그 때, 각 구현체의 예외 메시지를 다르게 설정하거나, 개별적인 변화를 별도의 인터페이스로 분리하는 것이 더 적절할 것으로 판단했습니다.

> 예외 메시지의 중복으로 인해 한쪽의 변경으로 다른 쪽의 수정을 놓치는 케이스도 발생할까요? 
즉, 같은 이유로 변경할 다른 쪽의 예외 메시지가 있을까요?

피드백 주신 내용처럼, `여러 구현체에서 동일한 예외 메시지를 공유하다가 특정 구현체만 변경이 필요할 경우`, 인터페이스에서 정의된 메시지를 사용하지 않게 되는 점이 일종의 LSP 위반으로 여길 수 있다고 생각합니다. 다만, 위에서 말했듯  `여러 구현체에서 동일한 예외 메시지를 공유하다가 특정 구현체만 변경이 필요할 경우`에 이미 해당 구현체는 인터페이스의 본래 목적과 다르기에 별도의 인터페이스로 분리하는 것이 좋다고 생각했습니다.

---

현재 제 코드가 옳다고 생각하지 않습니다. 피드백을 고려했을 때, 각 방법에는 장/단점이 있기에, 명확한 선택이 어렵다고 생각합니다. 

또, 크루들과 의견을 나누었을 때, 굳이 인터페이스에서 공통적인 예외 메시지를 정의하지 않고, 구현체마다 특화된 예외 메시지를 사용하자는 의견도 있었는데, 저는 코드의 의존성 측면에서도 인터페이스에 예외 메시지를 정의하는 것이 더 좋을 수도 있다고 생각했습니다. 

제 결론이 오답으로 향하는 것이 아닌가? 우려됩니다. 제 답변에 피드백 해주실 수 있는지 여쭤보고 싶습니다.

verus-j replied:
딱히 정답이 있는 문제가 아니여서 오답에 대한 걱정은 하지 않으셔도 될 것 같습니다. 이미 충분히 각 방법의 장/단점을 알고 계셔서 현재 상황에 맞춰서 사용하면 될 것 같습니다. 개인적으로는 현재 미션은 예외 메시지의 중복에 대한 깊은 고민을 할 정도로 복잡하지 않는 것 같습니다. 그래서 오히려 여러 방법 중 명확히 좋다고 생각하는 선택지를 찾지 못하는 상황인 것 같습니다. 우선은 각 장단점을 알고 있으니 강산이 원하는 심플한 방법으로 진행하다가 추후 수정이 필요한 순간이 생기면 그 때 고민해도 될 것 같습니다~

m-a-king replied:
감사합니다. 말씀해주신대로, 복잡한 문제에서는 오히려 명확히 선택지의 장단점이 눈에 띌 것 같습니다. 
진행된 피드백과 함께 제 코드를 돌아보고, 인터페이스에 예외메시지가 존재할 때의 문제점을 한 번 더 생각해보겠습니다!

verus-j commented on src/main/java/janggi/piece/strategy/block/RequiredBlockCountStrategy.java:
```diff
@@ -0,0 +1,30 @@
+import janggi.Position;
+import janggi.Route;
+
+public class RequiredBlockCountStrategy implements BlockStrategy {
```
굳이 인터페이스를 만든 이유가 있을까요? 구현체가 하나인데 인터페이스를 분리한 이유가 있는지 궁금합니다.

m-a-king replied:
첫 번째 이유로 확장을 고려했기 때문입니다.
현재는 구현체가 하나지만, 기물의 이동 경로를 차단하는 로직은 전략적으로 다양하게 분리될 수 있습니다.
예를 들어 `2개 이상 4개 미만의 기물이 있어야 한다, 홀/짝수 개의 기물이 있어야 한다` 와 같은 조건을 유연하게 관리하기 위해 인터페이스를 도입했습니다.

두 번째 이유는 구현체가 아닌 추상체(인터페이스)에 의존하는 구조를 만들고 싶었습니다.
의존 역전 원칙의 관점에서도 추상화에 의존하는 방향이 적절하다고 판단했습니다.

다만, 제 코드 전체를 돌아봤을 때 어떤 부분은 확장을 고려해 인터페이스를 분리했고, 어떤 부분은 그렇지 않아 설계에 일관성이 부족해 보일 수 있다는 점이 저 스스로도 모순처럼 느껴지기도 합니다.
또한 실제로 차단 전략이 추가될 가능성이 크지 않다는 점을 고려하면, 인터페이스 없이 구현체 하나로 단순하게 유지하는 것이 더 적절했을지 고민이 되어 이에 대해 의견을 여쭙고 싶습니다.

verus-j replied:
한국 장기는 1000년이 넘은 도메인인 것 같습니다. 새로운 차단 전략이 추가될까요?
한국 장기가 아닌 일본 장기를 구현하게 된다면 차단 전략이 바뀔 수도 있겠네요. 하지만, 현재 미션에서는 한국 장기만 다루고 있기에 지금 시점에 차단 전략에 대한 확장을 고민하지 않아도 될 것 같습니다. 그리고 인터페이스로 분리해뒀다고 차단 전략에 대해서는 확실히 확장성을 가지고 있다고 장담할 수 있을까요? 현재 구조에서는 구현체 하나로도 이후 필요할 경우 상황에 맞춘 인터페이스 추출로 유연하게 대처가능할 것 같습니다.

m-a-king replied:
> 인터페이스로 분리해뒀다고 차단 전략에 대해서는 확실히 확장성을 가지고 있다고 장담할 수 있을까요? 

확장의 범위는 항상 고민하고 있지만, 단정 짓기 어려운 문제라는 생각이 듭니다.
말씀을 듣고 보니, 저는 이 부분을 간과했던 것 같습니다. 어떤 방향으로 확장될지 불분명한 상황에서, 인터페이스로 분리해뒀다고 해서 확장 가능성이 보장된다고 생각한 건 성급했던 것 같습니다. 확실한 확장 방향 없이 구조를 미리 정하는 것이 오히려 혼란을 줄 수도 있다는 점을 다시 생각하게 됐습니다.

> 구현체 하나로도 이후 필요할 경우 상황에 맞춘 인터페이스 추출로 유연하게 대처가능할 것 같습니다.

구현체 하나로 충분히 동작하게 만든 뒤, 실제 확장이 필요해지는 시점에 그에 맞는 인터페이스를 추출하는 방향이 훨씬 현실적이고 유연한 접근이라는 생각이 들었습니다. 

---

올바르게 수정하겠습니다!

verus-j commented on src/main/java/janggi/piece/strategy/block/BlockStrategy.java:
```diff
@@ -0,0 +1,11 @@
+import janggi.Board;
+import janggi.Position;
+
+public interface BlockStrategy {
```
BlockStrategy는 어떤건가요? Block이 어떤 의미인가요?

m-a-king replied:
> BlockStrategy는 어떤건가요?

이동 경로에 무언가가 막고 있는 상황을 판단(검증)하는 역할입니다.

> Block이 어떤 의미인가요?

Block은 기물이 목표 지점까지 이동할 때, 경로상의 장애물을 의미합니다.

verus-j replied:
MoveStrategy와 합쳐보는건 어떨까요? 결국에는 MoveStrategy와 동일한 관심사라는 생각이 듭니다.

m-a-king replied:
혹시 이번에 변경된 코드에서도 그렇게 생각하는지 여쭤보고 싶습니다.

움직임의 모양을 검사하는 것(Move)과 경로의 장애물을 검사하는 것(Block)이 같은 관심사로 묶이는 이유가 내부 로직이 아니라 메서드의 파라미터로 결정될 수 있는 걸까요? 그렇다면 생각해보지 못한 좋은 인사이트같습니다!

verus-j replied:
처음에는 단순히 기물의 움직임에 대한 것이다 보니 관심사가 같다고 느꼈습니다. 그런데, 지금 다시 코드를 보니 관심사가 다른게 느껴지군요. 단순히 합쳤다가는 조합이 폭발하는 모양이 나올 수도 있겠네요. 그래서 현재는 MoveStrategy 내부에서 BlockStrategy를 사용하여 조합으로 풀어보면 어떨까라는 생각이 드네요.

```java
// 블록이 하나 필요
new StraightMoveStrategy(1) 
혹은
StraightMoveStrategy.withBlock(1)

// 블록이 있으면 안됨
new StraightMoveStrategy(0) 
혹은
StraightMoveStrategy.withNonBlock()
```
```java
public StraightMoveStrategy(int requireBlockCount) {
    this.requireBlockCountStrategy = new RequiredBlockCountStrategy(requireBlockCount);
}
```
위와 같이 작성해보는 것도 재밌을 것 같네요.

m-a-king replied:
너무 재밌을 것 같아요. 

> StraightMoveStrategy.withBlock(1)
StraightMoveStrategy.withNonBlock()

너무 좋습니다!

verus-j commented on src/main/java/janggi/piece/strategy/move/CurvedMoveStrategy.java:
```diff
@@ -0,0 +1,27 @@
+
+public class CurvedMoveStrategy implements MoveStrategy {
+
+    private final CurvedMovement movement;
```
CurvedMovement를 분리한 이유가 있을까요? CurvedMoveStrategy에서 CurvedMovement가 구현한 로직을 합쳐두면 안되나요? 동일한 관심사인데 불필요하게 분리되었다고 생각합니다

m-a-king replied:
맞습니다. CurvedMoveStrategy는 나중에 구현되었고, 당시에는 기존에 존재하던 CurvedMovement를 활용하려고 했던 것 같습니다.
말씀하신 대로 두 클래스는 동일한 관심사를 다루고 있음에도 불구하고 분리되어 있어, 오히려 응집도가 떨어지고 중복이 생겼습니다! 수정하겠습니다.

m-a-king replied:
```java
// 1
boolean isValidCurvedMovement = straightMovement == straightDistance && diagonalMovement == diagonalDistance;
        if (!isValidCurvedMovement) {
            throw new IllegalArgumentException(exceptionMessage);
        }
```

```java
// 2        
boolean isInvalidCurvedMovement = straightMovement != straightDistance || diagonalMovement != diagonalDistance;
        if (isInvalidCurvedMovement) {
            throw new IllegalArgumentException(exceptionMessage);
        }
```


```java
// 3
boolean isValidCurvedMovement = straightMovement == straightDistance && diagonalMovement == diagonalDistance;
        if (isValidCurvedMovement) {
            return;
        }
        throw new IllegalArgumentException(exceptionMessage);
```

어떤 코드가 더 가독성이 좋은지 여쭤보고 싶습니다.

verus-j replied:
어떤걸 택해도 크게 문제 없는 것 같습니다. 편한걸로 선택하시면 될 것 같습니다~

verus-j commented on src/main/java/janggi/piece/strategy/move/StraightMoveStrategy.java:
```diff
@@ -0,0 +1,22 @@
+        }
+    }
+
+    private boolean isNotStraightMovement(final int diffRow, final int diffColumn) {
+        // 대각선 이동 or 커브(마, 상) 이동
+        return Math.min(Math.abs(diffRow), Math.abs(diffColumn)) != 0;
+    }
```
isNotStraightMovement로 이미 주석의 내용을 담고 있지 않나요? 주석이 꼭 필요한 상황인가요?

m-a-king replied:
맞습니다. 메서드 명으로 조건을 명확하게 나타냈음에도 중복으로 주석을 추가했습니다. 삭제하겠습니다!

verus-j commented on src/main/java/janggi/Route.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Route {
```
Route는 객체가 아닌 것 같습니다. 유틸 클래스네요. 의도하신 걸까요?

m-a-king replied:
리팩토링 목록에 포함되어 있었습니다. 출발지와 도착지를 필드로 갖게 하고, 모든 스태틱 메서드를 인스턴스 메서드로 변경하겠습니다!

verus-j commented on src/main/java/janggi/Route.java:
```diff
@@ -0,0 +1,50 @@
+        route.removeLast();
+    }
+
+    private static void decideDirection(final List<Position> route,
+                                        final Position current,
+                                        final Position destination) {
+        int differenceOfRow = destination.subtractRow(current);
+        int differenceOfColumn = destination.subtractColumn(current);
+        int rowDirection = calculateDirection(differenceOfRow);
+        int columnDirection = calculateDirection(differenceOfColumn);
+        if (Math.abs(differenceOfRow) > Math.abs(differenceOfColumn)) {
+            route.add(current.adjust(rowDirection, 0));
+        }
+        if (Math.abs(differenceOfRow) < Math.abs(differenceOfColumn)) {
+            route.add(current.adjust(0, columnDirection));
+        }
+        if (Math.abs(differenceOfRow) == Math.abs(differenceOfColumn)) {
+            route.add(current.adjust(rowDirection, columnDirection));
+        }
+    }
```
direction은 diffeneceOfXXX의 음수 양수 여부로 정해지는군요. 한눈에 들어오지 않는 것 같습니다. 쉽게 이해할 수 있도록 개선해보면 어떨까요?

추가로 느낌상 decideDirection 한번당 route에 하나의 Position만 추가해야할 것 같은데 여러번 추가할 수 있지 않나요? 그럼 밖에서 for문으로 moveCount 만큼 Position을 더하는 로직을 이해하기 어렵지 않을까요?

m-a-king replied:
네 맞습니다. 반환 값으로 Position을 넘겨주는 것이 더 명확했을 것 같습니다.

좌표 VO를 통한 리팩토링을 진행하며 자연스럽게 해결했습니다!

verus-j commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,65 @@
+    protected abstract Piece createPiece(Position destination);
+
+    protected void validateMove(final Board board, final Position destination) {
+        validateIsAlly(board, destination);
+        validateMoveShape(board, destination);
+        validateIsBlock(board, destination);
+        validateSpecialRule(board, destination);
```
가능한 이동인지 확인하기 위해 단계를 나누셨군요! 잘하셨습니다. 궁금한게 SpecialRule은 어떤걸 의미하는걸까요? 개인적으로 특수 룰에 대한 정의가 애매한 것 같습니다. 왠지 위의 3가지 외에는 모든 특별하다고 생각되는 룰은 SpecialRule로 갈 것 같네요. 특별한 분류 없이 나머지 모든 규칙을 다루게 될 것 같네요. SpecialRule에 대해서 고민해보면 좋을 것 같습니다! 더 좋은 이름이 있을 수도 있고 위의 MoveShape이나 IsBlock에 포함되는 룰일 수도 있지 않을까요?

m-a-king replied:
맞습니다. 사실 MoveShape와 Block에 포함할 수 있습니다. 
하지만 MoveShape이나 Block에서 `졸/병의 본진 방향 움직임`이나 `포의 아군/적군의 모든 포를 넘을 수도, 잡을 수도 없다`는 규칙을 포함시키려니 오히려 더 복잡해진다고 느꼈습니다.

그래서 여러 기물들의 공통적인 특징이 아니라면, SpecialRule로 빼뒀습니다. 
결론은, 피드백을 주신 대로 SpecialRule, MoveShape, Block 등의 네이밍을 고민해 보는 것이 좋을 것 같습니다! 감사합니다.

verus-j commented on src/main/java/janggi/Position.java:
```diff
@@ -0,0 +1,78 @@
+    private static final int MAX_ROW = 10;
+    private static final int MAX_COLUMN = 9;
+
+    private static final Map<Integer, Map<Integer, Position>> CACHE;
+
+    static {
+        CACHE = new HashMap<>();
+        for (int row = MIN_ROW; row <= MAX_ROW; row++) {
+            CACHE.put(row, new HashMap<>());
+            for (int column = MIN_COLUMN; column <= MAX_COLUMN; column++) {
+                CACHE.get(row).put(column, new Position(row, column));
+            }
+        }
+    }
```
굳이 CACHE를 사용한 이유가 있을까요?

m-a-king replied:
위치 객체의 재사용이 잦기에 의미가 있을 것이라고 생각했는데, 어쩌면 크게 상관이 없을 것 같습니다. GC를 믿겠습니다!

verus-j commented on README.md:
```diff
@@ -1,3 +1,36 @@
 # java-janggi
```
> 현재 Position 클래스 외에도 거리나 행과 열의 개념을 명확히 표현하기 위해 추가적으로 Distance, Row/Column 을 값 객체나 ENUM으로 활용하여 개념을 포장하고자 합니다. 이 방식에 대한 더 나은 아이디어가 있는지 여쭤보고 싶습니다.

스스로 판단하기에 필요하다고 생각되면 시도해보면 될 것 같습니다~

m-a-king replied:
도메인 모델 로직의 가독성을 위해서 필요하다고 판단하고, 시도했습니다. 감사합니다!

verus-j reviewed (changes requested):
안녕하세요 강산
1단계 진행하시느라 고생 많으셨습니다.
피드백 남겨드렸습니다 확인부탁드려요~
궁금한점은 언제든 슬랙으로 물어봐주세요!

m-a-king committed:
refactor: 메서드 명 개선

m-a-king committed:
feat: 좌표 연산 VO 추가 및 개선 (Position, Route, Vector, Distance, Row, Column)

- Position: 행(Row)과 열(Column)을 조합한 좌표
- Route: 출발지와 도착지를 기반으로 이동 경로를 계산
- Vector: 방향과 크기를 가지는 벡터, Position 간 이동 차이를 표현
- Distance: 방향 없는 거리 정보, 벡터의 절댓값 또는 이동량을 의미
- Row / Column: 각각 행, 열의 값

m-a-king committed:
refactor: 움직임 전략 개선, 테스트 추가

m-a-king committed:
refactor: 블로킹 전략 개선, 테스트 추가

m-a-king committed:
refactor: 기물 개선 및 코드 리뷰 반영

m-a-king committed:
refactor: 보드 개선

m-a-king committed:
refactor: 코드 리포맷 등

m-a-king committed:
feat: Turn 추가

m-a-king committed:
refactor: Team 네이밍 개선, 설명 추가

m-a-king committed:
refactor: 네이밍 개선

m-a-king committed:
feat: 승리 점수, 점수 더하기, 비교 추가

m-a-king committed:
feat: 기본 행, 열 생성 추가

m-a-king committed:
feat: Pieces 일급 컬렉션 추가

m-a-king committed:
feat: PieceType 설명 추가

m-a-king committed:
test: Board 기물 움직임 테스트 추가

m-a-king committed:
test: 테스트 개선

m-a-king committed:
feat: 뷰 추가

m-a-king committed:
feat: Piece Team getter 추가

m-a-king committed:
feat: Player 추가

m-a-king committed:
feat: 장기 게임, GameOver 예외 추가

m-a-king committed:
fix: 예외가 발생해도 턴이 변경되는 문제 해결

m-a-king committed:
feat: 뷰 개선

m-a-king committed:
feat: 자신의 기물만을 움직일 수 있도록 수정

m-a-king committed:
refactor: 코드 리포맷

m-a-king commented:
## 베루스, 피드백 감사합니다. 1단계 2차 리뷰 요청과 함께, PR 본문을 업데이트했습니다.

verus-j commented on src/main/java/janggi/Board.java:
```diff
@@ -0,0 +1,74 @@
+
+    private final Map<Position, Piece> positionToPiece;
+
+    private Board(final Map<Position, Piece> positionToPiece) {
+        this.positionToPiece = positionToPiece;
+    }
+
+    public static Board from(final Pieces pieces) {
+        HashMap<Position, Piece> positionToPiece = new HashMap<>();
+
+        pieces.getPieces().forEach(piece ->
+                positionToPiece.put(piece.getPosition(), piece));
+
+        return new Board(positionToPiece);
+    }
```
https://github.com/woowacourse/java-janggi/pull/67#discussion_r2007610275

추가 답변 남겨봅니다. 생성자는 객체의 초기화만 진행하여 최대한 빠르게 동작해야한다고 생각하는 분들도 있지만, 개인적으로는 그러한 생성자를 만들기 위해서 정적 팩터리 메서드를 추가하는건 과하지 않나라고 생각합니다.

아래의 코드를 보고 간단하게 생성자로 작성할 수 있는데, 굳이 정적 팩터리 메서드로 Map을 변환하는 과정을 분리해야할 지 고민해보면 좋을 것 같습니다. 

```suggestion
    public Board(final Pieces pieces) {
        this.positionToPiece = pieces.getPieces()
                .stream()
                .collect(
                        Collectors.toMap(
                                Piece::getPosition,
                                piece -> piece
                        )
                );
    }
```

> 하지만 현재 대부분의 클래스가 정적 팩터리 메서드를 사용하고 있고, 그 흐름 안에서 이 클래스만 new를 사용하는 건 일관성 측면에서 오히려 눈에 띄고 다르게 느껴질 수 있습니다.

정말 대부분의 클래스가 정적 팩터리 메서드가 필요한지도 같이 고민해보면 좋을 것 같아요. 불필요한 정적 팩터리 메서드인데, 일관성 측면만 고려하여 나오게된 결과물이 아닌지도 생각해보면 좋을 것 같아요!

남겨 놓은 피드백은 완벽한 정답을 찾는 과정이 아니고 스스로 스타일을 찾아가는 시간이라고 생각하면 좋을 것 같아요~

m-a-king replied:
감사합니다! 사실 정적 팩터리 메서드를 활용할 명확한 사례를 보고 판단하면 쉬운 일이지만, 실제로 코드를 작성할 때 항상 엄격하게 고민할 만큼 중요한 문제인지 생각해 봤을 때, 그렇지 않다고 판단했기에 고민보다는 일관성을 지키는 쪽을 택한 것 같습니다.

그리고 말씀드리지 않았던 저의 기준이 하나 더 있습니다. `필드의 타입과 생성자의 파라미터 타입을 일치시키는 것입니다.` 그렇기에 필드가 Map, 생성을 위해서 전달받는 인자가 List라면 정적 팩토리 메서드를 활용했던 것 같습니다.

verus-j replied:
정적 팩터리 메서드는 충분히 얘기를 나눈 것 같네요.
추가로 아래의 내용만 더 얘기를 나눠보고 싶네요.

> 필드의 타입과 생성자의 파라미터 타입을 일치시키는 것입니다.

필드 타입과 생성자의 파라미터 타입을 일치시키게 되면 외부에서 객체의 내부 자료구조를 다 알게 되는거 아닌가요??

m-a-king replied:
> 필드 타입과 생성자의 파라미터 타입을 일치시키게 되면 외부에서 객체의 내부 자료구조를 다 알게 되는거 아닌가요??

맞습니다! 생성자의 파라미터 타입을 일치시키면, 외부에서 객체의 내부 자료구조를 모두 알게 됩니다. 간단한 값 객체가 아닌 경우에 이런 상황을 피하고 싶었습니다. 그렇기에 정적 팩토리 메서드를 사용했습니다.

생성자의 파라미터 타입을 일치시키지 않는다면 아래 예시와 같이 상속을 받은 경우, 생성자의 바디를 읽어야지만 어떤 필드를 가지는지 확인할 수 있다고 생각했습니다. 그렇기에 생성자의 파라미터 타입을 일치시키면 필드를 명확히 확인할 수 없는 상태(상속)에서, 생성자의 파라미터가 필드의 값과 같음을 암시할 수 있습니다.

또, 일관성을 위해서 정적 메서드를 사용하자는 것을 컨벤션으로 약속한다면, new 키워드가 나올 때 이것이 특별한 객체거나, 외부 라이브러리임을 암시한다고 생각했습니다. 물론 이것은 많은 경험에서 나온 기준이 아니기에, 잘못된 기준일 수 있다고 생각합니다.

```java
public class Cannon extends Piece {

    // 상위 클래스나 바디를 확인하지 않아도 나는 파라미터가 필드와 일치한다고 믿을 수 있다.
    private Cannon(final Position position,
                   final Team team,
                   final MoveStrategy moveStrategy,
                   final RequiredBlockCountStrategy blockStrategy) {
        super(position, team, moveStrategy, blockStrategy);
    }
    ...
}
```

```java
public class Cannon extends Piece {

    // 바디를 확인해서 필드를 파악할 수 있다.
    public Cannon(final Position position,
                   final Team team) {
        super(position, team, new StraightMoveStrategy(), new RequiredBlockCountStrategy());
    }
    ...
}
```

verus-j replied:
> 맞습니다! 생성자의 파라미터 타입을 일치시키면, 외부에서 객체의 내부 자료구조를 모두 알게 됩니다. 간단한 값 객체가 아닌 경우에 이런 상황을 피하고 싶었습니다. 그렇기에 정적 팩토리 메서드를 사용했습니다.

정적 팩터리 메서드를 사용한 이유를 이제야 이해했네요. 파라미터를 동일하게 유지하여 상속을 사용했을 때 필드를 파악하기 쉬울걸로 생각하여 파라미터를 맞췄다는 것도 이해했습니다.

가독성을 위해 좋은 시도일 수는 있으나, 한가지 의견을 하나 드리면 강산이 생각한 컨벤션을 유지하는 것은 어려울 수도 있습니다. 컴파일 에러도 나오지 않으며, 코드 리뷰와 같은 시간에 누군가 눈으로 일일이 확인하여야 컨벤션을 유지할 수 있기 때문입니다. 또한, 처음 누군가가 코드를 살펴본다면 하위 타입에서 super로 초기화하는 값들이 부모 타입의 필드임을 쉽게 알지 못할 수도 있습니다. 결국 계속해서 코드를 살피다가 어느 순간 같은 필드임을 깨닫게 되겠죠. 하지만, 그 타이밍에 누군가가 컨벤션을 놓치는 순간이 발생하면 다시 그러한 약속은 믿지 못하게 되어 부모 타입을 함께 살펴보게 될 수도 있습니다.

강산의 생각이 잘못된 건 아닙니다. 정말 필요한 컨벤션이라면 팀원 모두가 지킬 수 있으니깐요. 하지만, 누군가 반대되는 생각을 가지고 놓치기 시작한다면 적용하기 어려운 컨벤션이 될 수도 있습니다. 이번 미션에서는 페어와 같은 팀을 이뤄 진행했으니, 설명해주신 컨벤션을 서로 지키는 걸로 생각하겠습니다!

m-a-king replied:
베루스의 관점에 이어 실무적인 요소까지 짚어주셔서 감사합니다. 말씀해주신 내용 대부분에 공감합니다.
저 역시 정적 팩터리 메서드를 무조건 선호한다기보다는, 사용하다 보니 자연스럽게 익숙해진 정도였고, 만약 팀의 컨벤션이나 현실적인 여건상 맞추기 어렵다면 그에 맞춰 유연하게 대응할 수 있을 것 같습니다.

앞으로 비슷한 상황에서 베루스의 피드백이 떠오를 것 같네요. 감사합니다 ☺️

verus-j commented on src/main/java/janggi/Players.java:
```diff
@@ -0,0 +1,39 @@
+
+import janggi.piece.Pieces;
+
+public class Players {
```
Players에 대한 테스트가 없네요 ㅠㅠ

m-a-king replied:
아니 이런 실수를 ㅠㅠ 죄송합니다. 즉시 작성하겠습니다.

verus-j commented on src/main/java/janggi/Board.java:
```diff
@@ -0,0 +1,74 @@
+        if (currentTurnPlayer.getTeam() == piece.getTeam()) {
+            return;
+        }
+        throw new IllegalArgumentException("자신의 기물만을 움직일 수 있습니다.");
```
자신의 기물이 아닌 경우의 테스트 케이스가 없는 것 같습니다!

m-a-king replied:
감사합니다. 추가하겠습니다!

verus-j commented on src/main/java/janggi/Janggi.java:
```diff
@@ -0,0 +1,65 @@
+    public void play() {
+        Players players = Players.create();
+
+        Turn turn = Turn.create(1);
```
숫자 1은 매직 넘버 같군요. 그런데 그걸 위해서 저 값을 상수화 하는 것도 아쉬운 것 같습니다. 코드를 유심히 살펴보니 `Turn.create(1)`은 게임의 제일 처음 시작 턴을 만드는 것 같습니다. 이 의미를 잘담은 팩터리 메서드를 만들면 어떨까요? 그리고 팩터리 메서드 내부에서 숫자 1을 사용해 Turn을 생성하여 클라이언트 코드는 1의 존재를 모르게 하면 어떨까요?

m-a-king replied:
너무 좋습니다. 다른 곳에서는 관성적으로 사용하고, 정작 정적 팩터리 메서드를 활용할 좋은 예시에는 적용하지 못했었네요. 감사합니다! 수정하겠습니다.

verus-j commented on src/main/java/janggi/Janggi.java:
```diff
@@ -0,0 +1,65 @@
+        Players players = Players.create();
+
+        Turn turn = Turn.create(1);
+        Board board = Board.from(players.getBothPieces());
```
드는 생각이 getBothPieces가 반환하는 객체로 외부에서 Board를 생성하는 것 보다 Players에서 Board 객체를 생성하는 createBoard와 같은 팩터리 메서드를 추가하는게 어떨까요? Players는 Board를 생성하기에 필요한 정보를 가지고 있는 정보전문가 같습니다.

m-a-king replied:
Janggi 클래스의 play 메서드에서 Board 생성이 어색하다고 생각했는데, 말씀해 주신 부분이 근본적인 원인이었던 것 같습니다. 감사합니다! 수정하겠습니다.

verus-j commented on src/main/java/janggi/Janggi.java:
```diff
@@ -0,0 +1,65 @@
+
+        while (true) {
+            outputView.displayBoard(board);
+            outputView.displayScore(players);
```
2단계도 한번에 하셨군요. 피드백을 어떻게 진행하면 될까요? 한번에 봐드리고 2단계에서는 DB 적용만 보면 될까요?

m-a-king replied:
마음이 앞서서 1단계 수정과 함께 2단계 기능 요구 사항의 일부를 함께 진행한 것 같습니다. 

<img width="465" alt="image" src="https://github.com/user-attachments/assets/3707c49c-6cd0-40b2-9e6e-ec68f15c07e2" />

점수 계산은 현재 브랜치에 포함되어 있으니, 궁성은 2단계에서 작업하도록 하겠습니다. 
피드백은 베루스가 편한 대로 해주시면 좋을 것 같습니다!

verus-j commented on src/main/java/janggi/Janggi.java:
```diff
@@ -0,0 +1,65 @@
+            outputView.displayScore(players);
+
+            try {
+                Player player = players.getPlayer(turn.getCurrentTeam());
```
Players와 Turn을 합치는 것도 좋을 것 같네요. Players로 현재 턴의 플레이어를 찾는 동작이 한번에 가능했으면 클라이언트 코드가 알아야할 정보를 줄일 수 있을 것 같습니다.

m-a-king replied:
감사합니다. Players가 가진 정보를 바탕으로 
>드는 생각이 getBothPieces가 반환하는 객체로 외부에서 Board를 생성하는 것 보다 Players에서 Board 객체를 생성하는 createBoard와 같은 팩터리 메서드를 추가하는게 어떨까요? Players는 Board를 생성하기에 필요한 정보를 가지고 있는 정보전문가 같습니다.

위 피드백과 함께 수정하겠습니다.

verus-j commented on src/main/java/janggi/Board.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    public void movePiece(final Player player, final Position departure, final Position destination) {
+        Piece allyPiece = getPiece(departure);
```
allyPiece인지는 아직 위치만으로는 모르지 않나요?? selectedPiece가 더 맞을 것 같습니다.

m-a-king replied:
맞습니다. player가 추가되며 검증이 추가됐다면, 변수 명도 함께 수정했어야 했습니다. 짚어주셔서 감사합니다. 훨씬 명확한 것 같습니다.

verus-j commented on src/main/java/janggi/Board.java:
```diff
@@ -0,0 +1,74 @@
+
+    private void updateScore(final Player player, final Position destination) {
+        if (isExists(destination)) {
+            Score score = positionToPiece.remove(destination).die();
```
die라는 이름으로는 Score를 반환한다는걸 파악하기 어려울 것 같습니다. 심플하게 getScore()와 같은 이름으로 변경하는 건 어떨까요? 이미 remove를 통해서 die라는 의미를 이해할 것 같습니다.

그리고 updateScore에서 기물이 제거되는걸 예상할 수 있을까요? 아래의 updateBoard에서 remove를 호출하는게 맞는 것 같습니다.

m-a-king replied:
> die라는 이름으로는 Score를 반환한다는걸 파악하기 어려울 것 같습니다. 심플하게 getScore()와 같은 이름으로 변경하는 건 어떨까요? 이미 remove를 통해서 die라는 의미를 이해할 것 같습니다.

객체의 의인화에 너무 집중한 것일까요? 저는 기물이 잡힌다면, 외부에서 삭제하는 것보다 기물이 직접 메시지를 받고 스스로를 제거하는 것이 더 객체지향적이라고 생각했습니다.

차라리 get을 통해 찾아낸 기물에게 die를 진행할 수 있는 방법을 제공하는 방식이 나을까요? 즉, Board의 remove 메서드를 함수형 인터페이스로 전달받아, 기물이 스스로를 맵에서 삭제하는 방식도 괜찮을지 고민됩니다.

> 그리고 updateScore에서 기물이 제거되는걸 예상할 수 있을까요? 아래의 updateBoard에서 remove를 호출하는게 맞는 것 같습니다.

위에서 말씀드린 방식을 적용하면, 오히려 updateScore 메서드에서 기물이 제거되는 상황을 예측하기 더 어려워질 수도 있을 것 같습니다. 제 방법이 객체를 지나치게 자율적으로 설계했고, 오히려 가독성을 해치는 예시일까요?

m-a-king replied:
> Board의 remove 메서드를 함수형 인터페이스로 전달받아, 기물이 스스로를 맵에서 삭제하는 방식도 괜찮을지 고민됩니다.

적용해보니 기물(Piece)의 책임이 점점 무거워지고, 사이드 이펙트 발생 가능성도 우려되어 Board에서 기물을 제거하는 방식이 더 적절하다고 판단했습니다. 다만, 베루스의 의견도 여쭤보고 싶어 우선은 위 답변에서 제안된 구조대로 구현해두었습니다.

verus-j replied:
여기서는 기물을 제거하는 역할을 누가 가질지 고민해보면 좋을 것 같네요. 말한대로 기물이 본인 스스로를 보드에서 제거하는게 적절한 구현이라면 die가 올바를 수 있습니다. 하지만, 현재 상황에서는 보드가 기물을 제거하는게 더 명확하고, 쉽게 이해할 수 있는 코드인 것 같습니다. 따라서 의인화도 중요하지만 어떤 객체가 기물을 제거하는게 더 자연스럽고 좋은지를 살펴본 후 의인화를 고민해보면 좋을 것 같습니다.

추가로 의인화를 왜 사용하는 이유는 결국 캡슐화를 위해서라고 생각합니다. 불필요한 캡슐화를 위해 의인화를 사용하는게 올바른지 고민하는 것도 좋은 방향이 될 것 같네요~

m-a-king replied:
> 기물을 제거하는 역할을 누가 가질지 고민
더 명확하고, 쉽게 이해할 수 있는 코드
의인화도 중요하지만 어떤 객체가 기물을 제거하는게 더 자연스럽고 좋은지를 살펴본 후 의인화를 고민
불필요한 캡슐화를 위해 의인화를 사용하는게 올바른지 고민

역할 배분, 명확하고 쉽게 이해할 수 있는 코드, 캡슐화의 관점에서 제 코드를 수정한다면 보드에게로 기물 삭제의 책임이 이동할 것 같습니다. 감사합니다!

verus-j reviewed (changes requested):
안녕하세요 강산~
피드백 남겨놨습니다 확인부탁드려요~
질문 답변은 내일 저녁에 남겨둘게요!

m-a-king committed:
test: 기물 초기화 테스트 추가

m-a-king committed:
refactor: 빈 구현이 잦은 추상 메서드를 훅메서드로 변경

m-a-king committed:
refactor: 구현체가 하나이므로 인터페이스 삭제

m-a-king committed:
test: Players 테스트 추가

m-a-king committed:
refactor: Turn 정적 팩토리 메서드 개선

m-a-king committed:
refactor: 보드 생성 호출의 이동

- players는 Board를 생성하기에 필요한 정보를 가지고 있는 정보전문가

m-a-king committed:
test: 상대방의 기물을 움직일 수 없는 테스트 추가

m-a-king committed:
feat: Players가 Turn을 갖고 Board 초기화를 담당하도록 변경

- 클라이언트 코드가 알아야할 정보를 줄임.

m-a-king committed:
feat: 변수명 개선

- 위치만으로는 allyPiece인지 확신할 수 없음.

m-a-king committed:
refactor: 기물이 스스로를 보드에서 삭제하도록 수정, 로직 개선

m-a-king committed:
test: Players 보드 생성 테스트 추가

m-a-king commented on src/main/java/janggi/piece/strategy/move/SingleMoveStrategy.java:
```diff
@@ -0,0 +1,18 @@
+        if (Distance.of(departure, destination).getTotal() == SINGLE_STEP) {
+            return;
+        }
+        throw new IllegalArgumentException(exceptionMessage);
```
귀여운 토끼, 대문자 좋아

m-a-king commented on src/test/java/janggi/piece/SoldierTest.java:
```diff
@@ -0,0 +1,99 @@
+                .hasMessageContaining("본진을 향할 수 없습니다");
+    }
+
+    @ParameterizedTest
+    @CsvSource(value = {"1, 1", "2, 0"})
+    @DisplayName("졸/병은 2칸 이상 움직일 수 없다")
+    void move(int rowDirection, int columnDirection) {
```
displayName

m-a-king commented on src/main/java/janggi/view/InputView.java:
```diff
@@ -0,0 +1,42 @@
+
+        String move = scanner.nextLine();
+
+        String[] departureAndDestination = move.trim().split(" ");
+        String[] departurePosition = departureAndDestination[0].split(",");
+        String[] destinationPosition = departureAndDestination[1].split(",");
+
```
좋은데요

m-a-king commented on src/main/java/janggi/view/InputView.java:
```diff
@@ -0,0 +1,42 @@
+
+public class InputView {
+
+    private final Scanner scanner = new Scanner(System.in);
```
귀여워

m-a-king commented on src/main/java/janggi/view/MoveCommand.java:
```diff
@@ -0,0 +1,29 @@
+                          int destinationRow,
+                          int destinationColumn) {
+
+    public static MoveCommand of(final String departureRow,
+                                 final String departureColumn,
+                                 final String destinationRow,
+                                 final String destinationColumn) {
+
+        return new MoveCommand(
+                Integer.parseInt(departureRow),
+                Integer.parseInt(departureColumn),
+                Integer.parseInt(destinationRow),
+                Integer.parseInt(destinationColumn));
+    }
```
좋은데요

m-a-king commented on src/main/java/janggi/piece/Horse.java:
```diff
@@ -0,0 +1,42 @@
+
+import java.util.List;
+
+public class Horse extends Piece {
+
+    public static final int STRAIGHT_MOVEMENT = 1;
+    public static final int DIAGONAL_MOVEMENT = 1;
+
+    public Horse(final Position position, final Team team, final MoveStrategy moveStrategy, final RequiredBlockCountStrategy blockStrategy) {
+        super(position, team, moveStrategy, blockStrategy);
+    }
+
+    public static Horse of(Position position, Team team) {
+        return new Horse(position, team, new CurvedMoveStrategy(STRAIGHT_MOVEMENT, DIAGONAL_MOVEMENT), RequiredBlockCountStrategy.common());
+    }
+
+    public static List<Horse> defaultsOf(Team team) {
+        int defaultRow = Team.decideRow(1, team);
+        List<Integer> defaultColumns = List.of(3, 8);
+
+        return defaultColumns.stream()
+                .map(defaultColumn -> Horse.of(Position.of(defaultRow, defaultColumn), team))
+                .toList();
+    }
+
+    @Override
+    protected Piece createPiece(final Position position) {
+        return Horse.of(position, team);
+    }
+
+    @Override
+    public PieceType getType() {
+        return PieceType.HORSE;
+    }
+}
```
좋아요

m-a-king commented on src/main/java/janggi/piece/General.java:
```diff
@@ -0,0 +1,35 @@
+import janggi.piece.strategy.move.MoveStrategy;
+import janggi.piece.strategy.move.SingleMoveStrategy;
+
+public class General extends Piece {
+
+    public General(final Position position, final Team team, final MoveStrategy moveStrategy, final RequiredBlockCountStrategy blockStrategy) {
+        super(position, team, moveStrategy, blockStrategy);
+    }
```
다음 단계

m-a-king replied:
여기서 이러시면 곤란해요

m-a-king committed:
fix: CurvedMoveStrategy 생성 검증 논리 오류 해결

m-a-king commented on src/main/java/janggi/piece/strategy/move/CurvedMoveStrategy.java:
```diff
@@ -0,0 +1,34 @@
+        this.diagonalMovement = diagonalMovement;
+    }
+
+    private void validateCurvedMovement(final int straightMovement, final int diagonalMovement) {
+        if (straightMovement == 0 && diagonalMovement == 0) {
+            throw new IllegalArgumentException("직선과 대각선으로 모두 움직여야합니다.");
+        }
+    }
```
이건 진짜 좋았어요

m-a-king commented on src/main/java/janggi/piece/strategy/move/CurvedMoveStrategy.java:
```diff
@@ -0,0 +1,34 @@
+        if (distance.getStraight() == straightMovement && distance.getDiagonal() == diagonalMovement) {
+            return;
+        }
+        throw new IllegalArgumentException(exceptionMessage);
```
귀여워

m-a-king commented on src/test/java/janggi/coordinate/RouteTest.java:
```diff
@@ -0,0 +1,131 @@
+        Position destination = Position.of(4, 3);
+
+        //when
+        List<Position> expected = Route.of(departure, destination).calculate();
+
+        //then
+        List<Position> actual = List.of(Position.of(2, 3), Position.of(3, 3));
+        assertThat(expected).containsExactlyElementsOf(actual);
```
좋은데요

m-a-king commented on src/main/java/janggi/coordinate/Route.java:
```diff
@@ -0,0 +1,59 @@
+    private final Position departure;
+    private final Position destination;
+
+    private Route(final Position departure, final Position destination) {
+        this.departure = departure;
+        this.destination = destination;
+    }
+
+    public static Route of(final Position departure, final Position destination) {
+        return new Route(departure, destination);
+    }
```
흠......

m-a-king commented on src/main/java/janggi/piece/strategy/block/RequiredBlockCountStrategy.java:
```diff
@@ -0,0 +1,30 @@
+import janggi.coordinate.Position;
+import janggi.coordinate.Route;
+
+public class RequiredBlockCountStrategy {
+
+    private final int requireBlockCount;
+
+    public RequiredBlockCountStrategy(final int requireBlockCount) {
+        this.requireBlockCount = requireBlockCount;
+    }
```
많이 좋은데요

m-a-king committed:
feat: RequiredBlockCountStrategy 생성시, 블록 개수 음수 검증 추가

m-a-king committed:
refactor: 간단한 검증 추가, 코드 스타일 개선

verus-j commented on src/main/java/janggi/coordinate/Distance.java:
```diff
@@ -0,0 +1,60 @@
+        return Math.min(rowDistance, columnDistance);
+    }
+
+    public boolean isVertical() {
+        return rowDistance != 0 && columnDistance == 0;
+    }
+
+    public boolean isHorizontal() {
+        return rowDistance == 0 && columnDistance != 0;
+    }
+
+    public boolean isStraight() {
+        return isVertical() || isHorizontal();
+    }
```
isVertical과 isHorizontal은 isStraight와 테스트 코드에서만 사용하고 있군요. private으로 변경하고 테스트에서도 isStraight을 통해 검증하도록 수정하면 어떨까요?

m-a-king replied:
말씀하신대로 수정하니 의도가 더 명확해지는 것 같습니다. 수정하겠습니다!

verus-j commented on src/main/java/janggi/coordinate/Route.java:
```diff
@@ -0,0 +1,59 @@
+        return new Route(departure, destination);
+    }
+
+    public List<Position> calculate() {
+        return calculate(true);
+    }
+
+    public List<Position> calculate(boolean excludeDestination) {
```
```suggestion
    public List<Position> calculateIncludeDestination() {
        return calculate(true);
    }

    public List<Position> calculateExcludeDestination() {
        return calculate(false);
    }
    
    private List<Position> calculate(boolean excludeDestination) {
```
이렇게도 작성해볼 수 있을 것 같네요

m-a-king replied:
감사합니다! 궁성 구현에서 출발지도 포함되어야하는 경우가 생겨서 이런 방법으로 리팩토링해뒀었는데, 훨씬 명확하다고 느꼈습니다.

verus-j commented on src/main/java/janggi/coordinate/Route.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public List<Position> calculate(boolean excludeDestination) {
+        List<Position> route = new ArrayList<>();
+        route.add(departure);
+
+        Vector origin = departure.vectorTo(destination);
+        Vector straightPart = origin.extractStraightForDiagonal();
+        Vector diagonalPart = origin.subtract(straightPart);
+
+        for (Vector step : straightPart.splitToUnitVectors()) {
+            Position next = route.getLast().add(step);
+            route.add(next);
+        }
+
+        for (Vector step : diagonalPart.splitToUnitVectors()) {
+            Position next = route.getLast().add(step);
+            route.add(next);
+        }
```
각 반복문을 메서드 추출할 수 있을 것 같아요~

m-a-king replied:
감사합니다. 가독성을 위해서 메서드 추출을 진행하겠습니다!

verus-j reviewed (approved):
안녕하세요 강산
1단계 진행하시느라 고생많으셨습니다.
이번 단계는 여기서 이만 머지하겠습니다.
몇가지 간단한 피드백 남겨뒀으니 확인부탁드려요~
